src/Development/IDE/Core/Rules.hs

```diff
@@ -707,6 +743,17 @@ getModIfaceFromDiskRule = defineEarlyCutoff $ \GetModIfaceFromDisk f -> do
         case r of
             (diags, Just x) -> do
                 let fp = Just (hiFileFingerPrint x)
+                -- Check state of hiedb
+                se@ShakeExtras{hiedb,hiedbWriter} <- getShakeExtras
+                mrow <- liftIO $ HieDb.lookupHieFileFromSource hiedb (fromNormalizedFilePath f)
+                case mrow of
+                  Just row | ms_hs_date ms <= HieDb.modInfoTime (HieDb.hieModInfo row) -> pure ()
+                  _ -> liftIO $ do
+                    mhf <- runIdeAction "GetModIfaceFromDisk" se $ runMaybeT $
+                      readHieFileFromDisk (ml_hie_file $ ms_location ms)
+                    case mhf of
+                      Nothing -> pure ()
+                      Just hf -> indexHieFile hiedbWriter ms f hf (mi_decl_docs $ hirModIface x)
```

Comment:
Could you expand on this comment? What is this side effect needed for? 

Consider extracting it to a helper function